File: android-2wix-whatsapp/app/src/main/java/ru/twowix/whatsapp_shell/push/PushPayload.kt
package ru.twowix.whatsapp_shell.push

/**
 * Контракт data payload для push.
 *
 * Поддерживаемые поля:
 * - chatId
 * - phone
 * - clientName
 * - preview
 * - unreadCount
 * - targetUrl
 * - messageId
 * - type
 */
data class PushPayload(
  val chatId: String? = null,
  val phone: String? = null,
  val clientName: String? = null,
  val preview: String? = null,
  val unreadCount: Int? = null,
  val targetUrl: String? = null,
  val messageId: String? = null,
  val type: String? = null,
) {
  val title: String
    get() = (clientName?.takeIf { it.isNotBlank() } ?: phone?.takeIf { it.isNotBlank() } ?: "Новое сообщение")

  companion object {
    fun fromDataMap(data: Map<String, String>): PushPayload {
      fun get(key: String): String? = data[key]?.trim()?.takeIf { it.isNotEmpty() }
      fun getInt(key: String): Int? = get(key)?.toIntOrNull()

      return PushPayload(
        chatId = get("chatId"),
        phone = get("phone"),
        clientName = get("clientName"),
        preview = get("preview"),
        unreadCount = getInt("unreadCount"),
        targetUrl = get("targetUrl"),
        messageId = get("messageId"),
        type = get("type"),
      )
    }
  }

  fun bestTargetUrl(
    buildChatUrl: (String) -> String,
  ): String? {
    val direct = targetUrl?.trim()?.takeIf { it.isNotEmpty() }
    if (direct != null) return direct
    val cid = chatId?.trim()?.takeIf { it.isNotEmpty() } ?: return null
    return buildChatUrl(cid)
  }
}

File: ios-2wix-whatsapp/TwowixWhatsApp/PushPayload.swift
import Foundation

/// Контракт data payload push (как на Android).
struct PushPayload {
  let chatId: String?
  let phone: String?
  let clientName: String?
  let preview: String?
  let unreadCount: Int?
  let targetUrl: String?
  let messageId: String?
  let type: String?

  var title: String {
    let c = clientName?.trimmingCharacters(in: .whitespacesAndNewlines)
    let p = phone?.trimmingCharacters(in: .whitespacesAndNewlines)
    if let c, !c.isEmpty { return c }
    if let p, !p.isEmpty { return p }
    return "Новое сообщение"
  }

  static func from(userInfo: [AnyHashable: Any]) -> PushPayload {
    var merged: [AnyHashable: Any] = userInfo
    if let nested = userInfo["data"] as? [String: String] {
      for (k, v) in nested { merged[k] = v }
    }
    func s(_ key: String) -> String? {
      if let v = merged[key] as? String {
        let t = v.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.isEmpty ? nil : t
      }
      if let v = merged[key] as? Int {
        return String(v)
      }
      return nil
    }
    func i(_ key: String) -> Int? {
      if let v = merged[key] as? Int { return v }
      if let v = merged[key] as? String { return Int(v) }
      return nil
    }
    return PushPayload(
      chatId: s("chatId"),
      phone: s("phone"),
      clientName: s("clientName"),
      preview: s("preview"),
      unreadCount: i("unreadCount"),
      targetUrl: s("targetUrl"),
      messageId: s("messageId"),
      type: s("type")
    )
  }

  func bestTargetUrl(buildChatUrl: (String) -> String) -> String? {
    if let t = targetUrl?.trimmingCharacters(in: .whitespacesAndNewlines), !t.isEmpty {
      return t
    }
    guard let cid = chatId?.trimmingCharacters(in: .whitespacesAndNewlines), !cid.isEmpty else {
      return nil
    }
    return buildChatUrl(cid)
  }
}
